src/main/java/org/apache/ibatis/cache/decorators/SoftCache.java

```diff
@@ -87,7 +89,10 @@ public Object getObject(Object key) {
   @Override
   public Object removeObject(Object key) {
     removeGarbageCollectedItems();
-    return delegate.removeObject(key);
+    // See #2403 fix return value
+    @SuppressWarnings("unchecked") // assumed delegate cache is totally managed by this cache
+    SoftReference<Object> softReference = (SoftReference<Object>) delegate.removeObject(key);
+    return Optional.ofNullable(softReference).filter(reference -> Objects.nonNull(reference.get())).map(SoftReference::get).orElse(null);
```

Comment:
I know many people love Optional, but this looks overdoing.
Please do this in a classic way:

```java
return softReference == null ? null : softReference.get();
```

Reply:
Maybe we don't need to read the return value of the remove method, just like the API doc description 
```
/**
   * As of 3.3.0 this method is only called during a rollback
   * for any previous value that was missing in the cache.
   * This lets any blocking cache to release the lock that
   * may have previously put on the key.
   * A blocking cache puts a lock when a value is null
   * and releases it when the value is back again.
   * This way other threads will wait for the value to be
   * available instead of hitting the database.
   *
   *
   * @param key
   *          The key
   * @return Not used
   */
Object removeObject(Object key);
```